peps/pep-0788.rst

```diff
@@ -0,0 +1,633 @@
+
+        release_some_lock();
+        PyThreadState_Release();
+        Py_RETURN_NONE;
+    }
+
+Transitioning from old functions
+********************************
+
+The following code uses the old ``PyGILState`` APIs:
+
+.. code-block:: C
+
+    static int
+    thread_func(void *arg)
+    {
+        PyGILState_STATE gstate = PyGILState_Ensure();
+        /* It's not an issue in this example, but we just attached
+           a thread state for the main interpreter. If my_method() was
+           originally called in a subinterpreter, then we would be unable
+           to safely interact with any objects from it. */
+        if (PyRun_SimpleString("print(42)") < 0) {
+            PyErr_Print();
+        }
+        PyGILState_Release(gstate);
+        return 0;
+    }
+
+    static PyObject *
+    my_method(PyObject *self, PyObject *unused)
+    {
+        PyThread_handle_t handle;
+        PyThead_indent_t indent;
+
+        if (PyThread_start_joinable_thread(thread_func, NULL, &ident, &handle) < 0) {
+            return NULL;
+        }
+        Py_BEGIN_ALLOW_THREADS;
+        PyThread_join_thread(handle);
+        Py_END_ALLOW_THREADS;
+        Py_RETURN_NONE;
+    }
+
+This is the same code, updated to use the new functions:
+
+.. code-block:: C
+
+    static int
+    thread_func(void *arg)
+    {
+        PyInterpreterState *interp = (PyInterpreterState *)arg;
+        if (PyThreadState_Ensure(interp) < 0) {
+            fputs("Cannot talk to Python", stderr);
+            return -1;
+        }
+        if (PyRun_SimpleString("print(42)") < 0) {
+            PyErr_Print();
+        }
+        PyThreadState_Release();
+        return 0;
+    }
+
+    static PyObject *
+    my_method(PyObject *self, PyObject *unused)
+    {
+        PyThread_handle_t handle;
+        PyThead_indent_t indent;
+
+        PyInterpreterState *interp = PyInterpreterState_Hold();
+        if (PyThread_start_joinable_thread(thread_func, interp, &ident, &handle) < 0) {
+            PyInterpreterState_Release(interp);
+            return NULL;
+        }
+        Py_BEGIN_ALLOW_THREADS
+        PyThread_join_thread(handle);
+        Py_END_ALLOW_THREADS
+        Py_RETURN_NONE;
+    }
+
+
+Reference Implementation
+========================
+
+A reference implementation of this PEP can be found
+`here <https://github.com/ZeroIntensity/cpython/tree/pep-788-impl>`_.
+
+Rejected Ideas
+==============
+
+Using an interpreter ID instead of a interpreter state
+------------------------------------------------------
+
+Some iterations of this API took an ``int64_t interp_id`` parameter instead of
+``PyInterpreterState *interp``, because interpreter IDs cannot be concurrently
+deleted and cause use-after-free violations. :c:func:`PyInterpreterState_Hold`
+fixes this issue anyway, but an interpreter ID does have the benefit of
+requiring less magic in the implementation, but has several downsides:
+
+1. Nearly all existing APIs already return a :c:type:`PyInterpreterState`
```

Comment:
Use bullet points if the order is not important.

https://asrfs.github.io/Tech-Writing-for-Techies/output/html/02-03-lists-instructions.html#_bullet_or_numbered